modules/internal/ChapelArray.chpl

```diff
@@ -1943,22 +1943,21 @@ module ChapelArray {
   }
 
   @unstable("casting an array to an array type is unstable due to being a new feature — please share any feedback you might have")
-  operator :(arr: [], type t: []) {
-    // Would like to write:
-    //   var a: t = [elem in arr] elem:t.eltType;
-    // but we don't support type queries on array types (yet)
-    // (see https://github.com/chapel-lang/chapel/issues/8543)
-    var a: t;
-
-    if !arr.isRectangular() || !a.isRectangular() {
+  operator :(rhs: [], type t: []) {
+    var res: t;
+
+    if !rhs.isRectangular() || !res.isRectangular() {
       compilerError("Casts between arrays only support rectangular arrays");
-    } else if arr.rank != a.rank {
+    } else if rhs.rank != res.rank {
       compilerError("Casts between arrays require matching ranks");
     } else {
-      a = [elem in arr] elem:a.eltType;
+      type resType = res.eltType;
+      forall (i,j) in zip(res.domain, rhs.domain) {
```

Comment:
When imagining more cases of casting like this, someone might do:

```chpl
var A = [1, 2, 3];
var B = A: [1..10] int(32);
```
Note that the domain we are trying to cast to here is strictly larger than A's original domain.

It might be useful to have a better error message here than `size mismatch in zippered iteration`, but it would have to be a runtime check.

Reply:
Or alternatively, we could even allow this?

Reply:
No, I don't think we should allow this, or at least, certainly not in a first draft.  I like the idea of having a check, which I'm only realizing now would've been good to have some tests of in the original PR as well.  I'll add the check and some tests of it.